### PR TITLE
New Async Validator

### PR DESCRIPTION
**Pull request by ivanleomk**
Adding a new async validator

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit d25f55766c96089656708740cb856c58353ae147  | 
|--------|--------|

### Summary:
Introduced asynchronous validation for Pydantic models with new decorators, context handling, methods, tests, and documentation updates.

**Key points**:
- Added `async_field_validator` and `async_model_validator` decorators in `instructor/validators.py` for async validation.
- Introduced `AsyncValidationContext` and `AsyncValidationError` in `instructor/validators.py`.
- Updated `OpenAISchema` in `instructor/function_calls.py` with methods: `get_async_validators`, `get_model_coroutines`, `model_async_validate`, `execute_field_validator`, and `execute_model_validator`.
- Modified `process_response_async` in `instructor/process_response.py` to handle async validation.
- Updated `retry_async` in `instructor/retry.py` to catch `AsyncValidationError`.
- Added tests in `tests/llm/test_openai/test_async_decorator.py` for async validator functionality.
- Updated documentation in `docs/concepts/reask_validation.md` to include examples of async validation.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by ivanleomk** [comment]
@jonastemplestein I've implemented a version of the async validator, would love to hear your thoughts on it.

See Documentation here : https://1586cbb1.instructor.pages.dev/concepts/reask_validation/#async-validation

File: instructor/decorators.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+from typing import Callable, Any, TypeVar
```
Comment (jxnl) [inline instructor/decorators.py:1]: lets rename this to validators 
then also bring it to the main namespace in __init__.py

File: docs/concepts/reask_validation.md [hunk 2]
```diff
@@ -193,6 +193,253 @@ except ValidationError as e:
...
+
+This is especially useful if you are making multiple api calls in parallel. This implementation is heavily inspired by [pydantic-async-validation](https://github.com/team23/pydantic-async-validation/tree/64e4190494141c881ade30963390f91c3286dafc).
+
+### Field Validators
+
+To use our Async Validators, you'll need to import in the `OpenAISchema` class that we expose in instructor. Here's a quick example of how we can validate multiple fields at the same time
+
+```python hl_lines="12"
+from instructor.decorators import async_field_validator
+from instructor.function_calls import OpenAISchema
```
Comment (jxnl) [inline docs/concepts/reask_validation.md:208]: why not base model? we should try to not expose this externally

Comment (jxnl) [reply docs/concepts/reask_validation.md:208]: would almost rather 

`from instructor import BaseModel`

Comment (jxnl) [reply docs/concepts/reask_validation.md:208]: or from `instructor import AsyncBaseModel`

Comment (ivanleomk) [reply docs/concepts/reask_validation.md:208]: Managed to get this working with the default pydantic base model! :)

File: instructor/decorators.py (hunk 1 above)
Comment (ivanleomk) [reply instructor/decorators.py]: Renamed and added to the main namespace

**Comment by ivanleomk** [comment]
Implemented the warning when a user uses the synchronous client @jonastemplestein. I'll need to think a bit about how to implement a field-level validator which re-runs when a specific validator returns a changed value.

Will need to change a significant amount of code to support that :<

File: docs/concepts/reask_validation.md [hunk 3]
```diff
@@ -193,6 +193,255 @@ except ValidationError as e:
...
+1.  You can set the fields to run the same validator on. We don't support wildcard characters (eg. '\*' ) at this moment but might in the future.
+
+### Model Validators
+
+We can also use a model validator to validate a combination of fields. In this case, we want to make sure that every field is in caps. This helps us to prevent errors.
+
+```python hl_lines="11"
+from pydantic import BaseModel
+import asyncio
+from instructor import from_openai,async_model_validator
```
Comment (jxnl) [inline docs/concepts/reask_validation.md:254]:
```suggestion
from instructor import from_openai, async_model_validator
```

**Comment by jxnl** [comment]
i think this just needs ruff on the docs. 

`pytest --update-examples`

**Comment by ivanleomk** [comment]
@jxnl have updated the repo and ran ruff on it

**Review by jxnl — APPROVED** [review]

File: instructor/function_calls.py [hunk 4]
```diff
@@ -295,14 +434,39 @@ def parse_json(
...
+def openai_schema_helper(cls: T) -> T:
+    origin = get_origin(cls)
+
+    if origin is list:
+        return list[openai_schema_helper(cls.__args__[0])]
+
+    if origin is Literal:
+        return cls
+
+    if issubclass(cls, (str, int, bool, float, bytes)):
```
Comment (michael-quinlan) [inline instructor/function_calls.py:446]: I think this code doesnt handle UnionTypes, for example a BaseModel with something like
```
  name: str | None
```
At this point if you call ```issubclass(cls, (str, int, bool, float, bytes))``` on. UnionType you get ```TypeError: issubclass() arg 1 must be a class```.


Our code started crashing and I bisected it to this PR. In particular if I revert openai_schema to 
```
shema = wraps(cls, updated=())(
        create_model(
            cls.__name__ if hasattr(cls, "__name__") else str(cls),
            __base__=(cls, OpenAISchema),
        )
    )
    return cast(OpenAISchema, shema) 
```
and bypass this whole new method then it works again.

FYI we use Hakiu and i haven't validated if this code works on OpenAi, however are they any tests that cover UnionTypes?

I can also see if I can come up with a fix, but those who know the entire code base better might be faster.

Comment (michael-quinlan) [reply instructor/function_calls.py:446]: So i think if you add this
```
    if isinstance(cls, UnionType):
        for arg in get_args(cls):
            return openai_schema_helper(arg)
```
on line 444 (i.e above ```if issubclass(cls, (str, int, bool, float, bytes)): ...'```) then it works.

You will also need to 
```
from typing import Annotated, Any, Optional, TypeVar, cast, get_origin, Literal, get_args
from types import UnionType
```

Comment (lazyhope) [reply instructor/function_calls.py:446]: Not only that, it doesn't work for types like dict[str, str] either. The current implementation really needs to be carefully reviewed and fixed. @jxnl @ivanleomk

**Pull request merged**